lib/learn_to_rank/explain_scores.rb

```diff
@@ -0,0 +1,54 @@
+module LearnToRank
+  class ExplainScores
+    # Given an explanation for a document, returns a hash of the total scores
+    # e.g. { title_score: 0.2, description_score: 0.2, ... }
+    # attr_reader :title_score, :description_score, :all_searchable_text_score, :indexable_content_score
+
+    def initialize(explain)
+      @scores = scores_from_explain(explain)
+      PERMITTED_FIELDS.each do |field|
+        define_singleton_method field do
+          scores[field]
+        end
+      end
+    end
+
+  private
+
+    attr_reader :scores
+
+    PERMITTED_FIELDS = [:title_score, :description_score, :all_searchable_text_score, :indexable_content_score]
+
+    def normalise(n)
+      "%.10f" % (((n || 0)  + 0.01) * 10000)
+    end
+
+    def field_permitted?(field)
+      PERMITTED_FIELDS.include? field
+    end
+
+    def scores_from_explain(explain)
+      description = explain["description"]
+      if description.include? "PerFieldSimilarity"
+        # is a bm25 score so return that value
+        field = (description.split("(")[1].split(":")[0].split(".")[0] + "_score").downcase.to_sym
+        return {} unless field_permitted? field
+
+        value = explain["value"]
+        { field => value }
+      elsif ["sum of:", "max of:", "function score, product of:"].include? description
+        # is not a bm25 score so return the sum of scores contained within
```

Comment:
Why the sum of the scores within and not the aggregate score Elasticsearch already gives us? (sum, max, etc(

Reply:
We want the summed scores for each field e.g. title, description. Without going into "function score, product of" or a "sum of: " we don't know whether the value is for a filter / different field. If we did use `sum of:` we might get the fields combined (e.g. title and description) but I think there is value in having the title score and description score as separate features. However, this is quite a hack which may even be broken. It'd be nice to prove that it works as expected and do some work to make it faster.

E.g. https://www.gov.uk/api/search.json?q=national%20insurance&fields=title&debug=explain&count=1